// @slack/bolt/dist/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var ErrorCode;
(function (ErrorCode) {
    ErrorCode["AppInitializationError"] = "slack_bolt_app_initialization_error";
    ErrorCode["AuthorizationError"] = "slack_bolt_authorization_error";
    ErrorCode["ContextMissingPropertyError"] = "slack_bolt_context_missing_property_error";
    ErrorCode["ReceiverAckTimeoutError"] = "slack_bolt_receiver_ack_timeout_error";
    ErrorCode["ExpressReceiverAuthenticityError"] = "slack_bolt_express_receiver_authenticity_error";
    /**
     * This value is used to assign to errors that occur inside the framework but do not have a code, to keep interfaces
     * in terms of CodedError.
     */
    ErrorCode["UnknownError"] = "slack_bolt_unknown_error";
})(ErrorCode = exports.ErrorCode || (exports.ErrorCode = {}));
function errorWithCode(message, code) {
    const error = new Error(message);
    error.code = code;
    return error;
}
exports.errorWithCode = errorWithCode;
function asCodedError(error) {
    if (error.code !== undefined) {
        return error;
    }
    error.code = ErrorCode.UnknownError;
    return error;
}
exports.asCodedError = asCodedError;
